### ISSUE TITLE
remove `unwrap`s from git2Index

### ISSUE THREAD

**Issue by danieleades** [Hirevo/alexandrie#51]
there's a lot of `unwrap`s in the git2 index.

Internally the `git2::Repository` is held in a mutex, which is unlocked and unwrapped in every method. It's not clear what purpose the Mutex serves in this module.

I suspect that the mutex is added so that the index itself is `Sync`, meaning it can be used "naked" in the tide::Server state. is that correct?

It might be better to move this into the calling code.

the application state could be wrapped (i'm guessing something like `State = Arc<Mutex<InternalState>>` before passing into the server, rather than the low-level bits. The result is the same, but you move the responsibility for ensuring thread-safe access to the application.

**Comment by Hirevo** [comment]
Yes, the `Mutex` in `Git2Index` was indeed to have it be `Sync`.  
At the time, I thought it would be fine but you made very good points about how that lock is not actually useful and how having it `Sync` is not accurate. It's entirely my bad on that.  
I am sold on removing the lock from there and using an `Arc<Mutex<config::State>>` as Tide's state.

**Comment by danieleades** [comment]
no problem! it also lifts the restriction that all the methods in the `Indexer` trait should take `&self`. Being able to take `&mut self` allows implementors of `Index` to avoid using interior mutability to modify themselves.
